source/common/router/scoped_config_impl.cc

```diff
@@ -3,6 +3,109 @@
 namespace Envoy {
 namespace Router {
 
+bool ScopeKey::operator!=(const ScopeKey& other) const { return !(*this == other); }
+
+bool ScopeKey::operator==(const ScopeKey& other) const {
+  if (this->fragments_.size() != other.fragments_.size()) {
+    return false;
+  }
+  if (this->fragments_.empty() || other.fragments_.empty()) {
```

Comment:
Do we need an explicit check here for size and empty? Won't the `std::equal` below cover these? You might need to use the variant where you provide `other.fragments_.end()`.

Reply:
so I didn't know if the std::equal impl will compare size first and then go over the two ranges and compare one by one, so had my own version of size comparison. Looked into the code and it does compare size first for random access iterator.
I will keep the either "empty equals to nothing" statement though, std::equal(empty, empty) evaluates to true.